### PR TITLE
[FLINK-16667][python][client] Support new Python dependency configuration options in flink-client.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request introduces new Python dependency configuration options in flink-client.*

## Brief change log

  - *Add logic to the constructor of `ProgramOptions` to parse the python command line arguments to configurations.*
  - *Write the python configurations to `ExecutionConfigAccessor` when constructing it from `ProgramOptions`.*

## Verifying this change

This change is already covered by existing tests, such as *PythonDependencyUtilsTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java [hunk 1]
```diff
@@ -22,43 +22,36 @@
...
 
-	private final String jarFilePath;
+	private String jarFilePath;
 
-	private final String entryPointClass;
+	protected String entryPointClass;
 
 	private final List<URL> classpaths;
 
-	private final String[] programArgs;
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java:61]: unnecessary change

File: flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		try {
+			ClassLoader classLoader;
+			try {
+				classLoader = new URLClassLoader(
+					new URL[]{PackagedProgramUtils.getPythonJar()},
+					Thread.currentThread().getContextClassLoader());
+			} catch (RuntimeException e) {
+				LOG.warn(
+					"An attempt to load the flink-python jar from the \"opt\" directory failed, " +
+						"fall back to use the context class loader to reflect the flink-python class.", e);
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java:67]: Remove `to reflect the flink-python class`.

File: flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+				classLoader = Thread.currentThread().getContextClassLoader();
+			}
+			Class<?> utilClazz = Class.forName(
+				"org.apache.flink.python.util.PythonDependencyUtils",
+				false,
+				classLoader);
+			Method utilMethod = utilClazz.getMethod(
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java:74]: utilMethod -> parseCommandLineMethod

File: flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	@Override
+	public void validate() throws CliArgsException {}
+
+	public static boolean isPython(CommandLine line) {
+		return line.hasOption(PY_OPTION.getOpt()) ||
+			line.hasOption(PYMODULE_OPTION.getOpt()) ||
+			"org.apache.flink.client.python.PythonGatewayServer".equals(line.getOptionValue(CLASS_OPTION.getOpt()));
+	}
+
+	public static boolean containsPythonDependency(CommandLine line) {
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java:130]: Rename to containsPythonDependencyOptions?

File: flink-python/src/main/java/org/apache/flink/python/util/PythonDependencyUtils.java [hunk 5]
```diff
@@ -77,6 +83,23 @@ public static Configuration configurePythonDependencies(
 		return pythonDependencyManager.getConfigWithPythonDependencyOptions();
 	}
 
+	public static Configuration parseCommandLine(CommandLine commandLine) {
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/python/util/PythonDependencyUtils.java:86]: Rename to parsePythonDependencyConfiguration

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments in the latest commit.

**Comment by dianfu** [comment]
@WeiZhong94  Thanks a lot for the update. 

@aljoscha This PR refactored the logic about Python into a separate PythonProgramOptions. It would be great if you could take a look at it. Thanks a lot!

File: flink-clients/src/main/java/org/apache/flink/client/cli/ExecutionConfigAccessor.java [hunk 6]
```diff
@@ -74,6 +74,11 @@ public static ExecutionConfigAccessor fromProgramOptions(final ProgramOptions op
 
 		SavepointRestoreSettings.toConfiguration(options.getSavepointRestoreSettings(), configuration);
 
+		// If the Python configuration exists, write them to the configuration for Python program/UDF execution.
+		if (options instanceof PythonProgramOptions) {
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ExecutionConfigAccessor.java:78]: Instead of the `instanceof` here, I think it's better to add a method to `ProgramOptions` for adding additional configuration. Something like `(addToConfiguration(Configuration)`. On the base `ProgramOptions` this would do nothing but in `PythonProgramOptions` this would add the additional python configuration.

If we really want to do it nicely, though, we can even add the current code in this method here to the new `ProgramOptions.addToConfiguration()`. This way we treat everything the same.

Comment (aljoscha) [reply flink-clients/src/main/java/org/apache/flink/client/cli/ExecutionConfigAccessor.java:78]: Ah, better name is probably `applyToConfiguration()`, and hand in a `Configuration`.

**Comment by WeiZhong94** [comment]
@aljoscha @dianfu  I have updated this PR according to your comments. Please take a look. :)

File: flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java [hunk 7]
```diff
@@ -184,10 +175,65 @@ public SavepointRestoreSettings getSavepointRestoreSettings() {
...
+
+		configuration.setBoolean(DeploymentOptions.ATTACHED, !getDetachedMode());
+		configuration.setBoolean(DeploymentOptions.SHUTDOWN_IF_ATTACHED, isShutdownOnAttachedExit());
+		ConfigUtils.encodeCollectionToConfig(configuration, PipelineOptions.CLASSPATHS, getClasspaths(), URL::toString);
+		SavepointRestoreSettings.toConfiguration(getSavepointRestoreSettings(), configuration);
+	}
+
+	public static ProgramOptions create(CommandLine line) throws CliArgsException {
+		if (isPython(line) || containsPythonDependencyOptions(line)) {
+			return createPythonPropramOptions(line);
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java:191]: typo: createPythonPropramOptions  -> createPythonProgramOptions

File: flink-python/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+ */
+public class PythonProgramOptions extends ProgramOptions {
+
+	private final Configuration pythonConfiguration;
+
+	private final boolean isPython;
+
+	public PythonProgramOptions(CommandLine line) throws CliArgsException {
+		super(line);
+		isPython = isPython(line);
```
Comment (dianfu) [inline flink-python/src/main/java/org/apache/flink/client/cli/PythonProgramOptions.java:47]: The name `isPython` is confusing. Considering the class `PythonProgramOptions` is specify for Python, should it always be true? I guess you mean `isEntryPointPython`? If so we should rename the variable `isPython` and the util method `isPython` to `isEntryPointPython`. What do you think?

File: flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java [hunk 9]
```diff
@@ -184,10 +175,65 @@ public SavepointRestoreSettings getSavepointRestoreSettings() {
...
+		} else {
+			return new ProgramOptions(line);
+		}
+	}
+
+	static boolean isPython(CommandLine line) {
```
Comment (dianfu) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java:197]: Should we move these kinds of util methods to a class such as `ProgramOptionsUtils` to make `ProgramOptions` more clean?  What's your thought? @aljoscha

File: flink-python/src/test/java/org/apache/flink/python/util/PythonDependencyUtilsTest.java [hunk 10]
```diff
@@ -184,20 +197,83 @@ public void testPythonExecutables() {
 		verifyConfiguration(expectedConfiguration, actual);
 	}
 
+	@Test
+	public void testCreateProgramOptionsWithPythonCommandLine() throws CliArgsException {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/python/util/PythonDependencyUtilsTest.java:201]: What about create a PythonProgramOptionsTest and move this test case there.

File: flink-python/src/test/java/org/apache/flink/python/util/PythonDependencyUtilsTest.java [hunk 11]
```diff
@@ -184,20 +197,83 @@ public void testPythonExecutables() {
...
+		assertEquals("a.txt#b_dir", config.get(PYTHON_REQUIREMENTS));
+		assertEquals("c.zip#venv,d.zip", config.get(PythonOptions.PYTHON_ARCHIVES));
+		assertEquals("bin/python", config.get(PYTHON_EXECUTABLE));
+		assertArrayEquals(
+			new String[] {"--python", "test.py", "--pyModule", "test", "userarg1", "userarg2"},
+			programOptions.getProgramArgs());
+	}
+
+	@Test
+	public void testCreateProgramOptionsWithLongOptions() throws CliArgsException {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/python/util/PythonDependencyUtilsTest.java:227]: ditto

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have updated this PR according to your comments.

**Comment by dianfu** [comment]
@WeiZhong94 Thanks a lot for the update. LGTM. 

@aljoscha Could you help to take a further look at this PR? Thanks a lot!

**Comment by WeiZhong94** [comment]
@flinkbot run travis

**Comment by WeiZhong94** [comment]
@flinkbot run azure

File: flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java [hunk 12]
```diff
@@ -152,6 +112,26 @@ public ProgramOptions(CommandLine line) throws CliArgsException {
...
+			args = Arrays.copyOfRange(args, 1, args.length);
+		}
+
+		return args;
+	}
+
+	public void validate() throws CliArgsException {
+		// Java program should be specified a JAR file
+		if (getJarFilePath() == null) {
+			throw new CliArgsException("Java program should be specified a JAR file.");
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java:131]: I think that's not always necessary, if the user jar is in the lib folder it's enough to specify an entrypoint Class.

Comment (WeiZhong94) [reply flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java:131]: Yes, either of entrypoint class and jar file is enough. But before this PR, we only check whether ProgramOptions.getJarFilePath() is null in `CliFrontend`, so I kept the original logic in the validate method. It is ok for me to change this validate logic in this PR. What do you think?
